### PR TITLE
[BEAM-614] Updates FileBasedSource to support CompressionTypes.AUTO.

### PR DESCRIPTION
**Pull request by chamikaramj**
Updates FileBasedSource to support CompressionTypes.AUTO.

Fixes some tests that were not properly being tested.

Adds tests for CompressionTypes.AUTO.

### CODE REVIEW

**Comment by chamikaramj** [comment]
R: @charlesccychen

File: sdks/python/apache_beam/io/filebasedsource_test.py [hunk 1]
```diff
@@ -44,18 +45,21 @@ def read_records(self, file_name, range_tracker):
     f = self.open_file(file_name)
     try:
       start = range_tracker.start_position()
-      f.seek(start)
+      if start > 0:
+        f.seek(start)
```
Comment (charlesccychen) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: Can you combine this with the clause below and just do f.seek(start - 1)?  Please add a comment in that case.

File: sdks/python/apache_beam/io/filebasedsource.py [hunk 2]
```diff
@@ -72,21 +71,15 @@ def __init__(self,
...
     self._pattern = file_pattern
     self._concat_source = None
     self._min_bundle_size = min_bundle_size
     if not fileio.CompressionTypes.is_valid_compression_type(compression_type):
       raise TypeError('compression_type must be CompressionType object but '
                       'was %s' % type(compression_type))
     self._compression_type = compression_type
-    if compression_type != fileio.CompressionTypes.UNCOMPRESSED:
+    if (compression_type != fileio.CompressionTypes.UNCOMPRESSED and
+        compression_type != fileio.CompressionTypes.AUTO):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/filebasedsource.py]: Done.

File: sdks/python/apache_beam/io/filebasedsource_test.py (hunk 1 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: Done.

**Comment by chamikaramj** [comment]
R: @robertwb please merge.

**Comment by chamikaramj** [comment]
Oops, forgot to push updates. Done now.

File: sdks/python/apache_beam/io/filebasedsource_test.py [hunk 3]
```diff
@@ -44,18 +45,23 @@ def read_records(self, file_name, range_tracker):
     f = self.open_file(file_name)
     try:
       start = range_tracker.start_position()
-      f.seek(start)
       if start > 0:
-        f.seek(-1, os.SEEK_CUR)
+        # Any line that starts after 'start' does not belong to the current
+        # bundle. Seeking to (start - 1) and skipping a line moves the current
+        # position of the starting position of the first line that belongs to
```
Comment (charlesccychen) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: nit: "position _to_ the starting position"

Comment (chamikaramj) [inline sdks/python/apache_beam/io/filebasedsource_test.py]: Done. Thanks.

**Comment by charlesccychen** [comment]
Thanks! LGTM.

**Comment by chamikaramj** [comment]
Thanks Robert.

**Comment by jiajiayb** [comment]
**Hello! When I trying the 'Evaluation and Batch Prediction' notebook in census tutorial, I met an error and have no idea why this is happening. So the script I am using is exactly the same as the one in the tutorial, which is as follows:**

```
import apache_beam as beam
import google.cloud.ml as ml
import google.cloud.ml.analysis as analysis
import google.cloud.ml.io as io
import json
import os

def extract_values((example, prediction)):
    import tensorflow as tf
    tf_example = tf.train.Example()
    tf_example.ParseFromString(example.values()[0])
    feature_map = tf_example.features.feature
    values = {'target': feature_map['target'].float_list.value[0]}
    values.update(prediction)
    return values

OUTPUT_DIR = '/content/datalab/tmp/ml/census/evaluate'
pipeline = beam.Pipeline('DirectPipelineRunner')

eval_features = (pipeline | 'ReadEval' >> io.LoadFeatures('/content/datalab/tmp/ml/census/preprocessed/features_eval*'))
trained_model = pipeline | 'LoadModel' >> io.LoadModel('/content/datalab/tmp/ml/census/model/model')
evaluations = (eval_features | 'Evaluate' >> ml.Evaluate(trained_model) |
beam.Map('ExtractEvaluationResults', extract_values))
eval_data_sink = beam.io.TextFileSink(os.path.join(OUTPUT_DIR, 'eval'), shard_name_template='')
evaluations | beam.io.textio.WriteToText(os.path.join(OUTPUT_DIR, 'eval'), shard_name_template='')

pipeline.run()
```

**The error I get is as follow:**

`ValueErrorTraceback (most recent call last)
in ()
36 # evaluation
37
---> 38 eval_features = (pipeline | 'ReadEval' >> io.LoadFeatures('/content/datalab/tmp/ml/census/preprocessed/features_eval*'))
39 trained_model = pipeline | 'LoadModel' >> io.LoadModel('/content/datalab/tmp/ml/census/model/model')
40 evaluations = (eval_features | 'Evaluate' >> ml.Evaluate(trained_model) |

/usr/local/lib/python2.7/dist-packages/apache_beam/transforms/ptransform.pyc in ror(self, pvalueish)
727
728 def ror(self, pvalueish):
--> 729 return self.transform.ror(pvalueish, self.label)
730
731 def apply(self, pvalue):

/usr/local/lib/python2.7/dist-packages/apache_beam/transforms/ptransform.pyc in ror(self, left, label)
435 pvalueish = _SetInputPValues().visit(pvalueish, replacements)
436 self.pipeline = p
--> 437 result = p.apply(self, pvalueish, label)
438 if deferred:
439 return result

/usr/local/lib/python2.7/dist-packages/apache_beam/pipeline.pyc in apply(self, transform, pvalueish, label)
207 try:
208 old_label, transform.label = transform.label, label
--> 209 return self.apply(transform, pvalueish)
210 finally:
211 transform.label = old_label

/usr/local/lib/python2.7/dist-packages/apache_beam/pipeline.pyc in apply(self, transform, pvalueish, label)
243 transform.type_check_inputs(pvalueish)
244
--> 245 pvalueish_result = self.runner.apply(transform, pvalueish)
246
247 if type_options is not None and type_options.pipeline_type_check:

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/runner.pyc in apply(self, transform, input)
145 m = getattr(self, 'apply_%s' % cls.name, None)
146 if m:
--> 147 return m(transform, input)
148 raise NotImplementedError(
149 'Execution of [%s] not implemented in runner %s.' % (transform, self))

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/runner.pyc in apply_PTransform(self, transform, input)
151 def apply_PTransform(self, transform, input):
152 # The base case of apply is to call the transform's apply.
--> 153 return transform.apply(input)
154
155 def run_transform(self, transform_node):

/usr/local/lib/python2.7/dist-packages/google/cloud/ml/io/transforms.pyc in apply(self, pvalue)
148 file_pattern=self._file_pattern,
149 coder=mlcoders.ExampleProtoCoder(),
--> 150 compression_type=self._compression_type))
151
152

/usr/local/lib/python2.7/dist-packages/apache_beam/transforms/ptransform.pyc in ror(self, left, label)
435 pvalueish = _SetInputPValues().visit(pvalueish, replacements)
436 self.pipeline = p
--> 437 result = p.apply(self, pvalueish, label)
438 if deferred:
439 return result

/usr/local/lib/python2.7/dist-packages/apache_beam/pipeline.pyc in apply(self, transform, pvalueish, label)
243 transform.type_check_inputs(pvalueish)
244
--> 245 pvalueish_result = self.runner.apply(transform, pvalueish)
246
247 if type_options is not None and type_options.pipeline_type_check:

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/runner.pyc in apply(self, transform, input)
145 m = getattr(self, 'apply_%s' % cls.name, None)
146 if m:
--> 147 return m(transform, input)
148 raise NotImplementedError(
149 'Execution of [%s] not implemented in runner %s.' % (transform, self))

/usr/local/lib/python2.7/dist-packages/apache_beam/runners/runner.pyc in apply_PTransform(self, transform, input)
151 def apply_PTransform(self, transform, input):
152 # The base case of apply is to call the transform's apply.
--> 153 return transform.apply(input)
154
155 def run_transform(self, transform_node):

/usr/local/lib/python2.7/dist-packages/google/cloud/ml/dataflow/io/tfrecordio.pyc in apply(self, pvalue)
164
165 def apply(self, pvalue):
--> 166 return pvalue.pipeline | beam.Read(_TFRecordSource(*self._args))
167
168

/usr/local/lib/python2.7/dist-packages/google/cloud/ml/dataflow/io/tfrecordio.pyc in init(self, file_pattern, coder, compression_type)
115 file_pattern=file_pattern,
116 compression_type=compression_type,
--> 117 splittable=False)
118 self._coder = coder
119

/usr/local/lib/python2.7/dist-packages/apache_beam/io/filebasedsource.pyc in init(self, file_pattern, min_bundle_size, compression_type, splittable)
74
75 if compression_type == fileio.CompressionTypes.AUTO:
---> 76 raise ValueError('FileBasedSource currently does not support '
77 'CompressionTypes.AUTO. Please explicitly specify the '
78 'compression type or use '

ValueError: FileBasedSource currently does not support CompressionTypes.AUTO. Please explicitly specify the compression type or use CompressionTypes.UNCOMPRESSED if file is uncompressed.`

**Could you please help me have a look what might be the issue for generating the error message. Thanks a lot!**

**Comment by charlesccychen** [comment]
@jiajiayb: the error message you are referring to was removed in approximately 10/2016 - 11/2016.  We believe you may be using an old version of the Google Distribution of the Apache Beam SDK.  Can you update to the latest version of the "google-cloud-dataflow" pip package or if this still occurs, post a copy of your "pip freeze" output?  Thanks.

**Comment by jiajiayb** [comment]
@charlesccychen I think you are right. When I run 'pip freeze' in google datalab notebook, it indicates 'google-cloud-dataflow==0.4.2', which I think it is an older version of apache beam. Do you by any chance know how to upgrade google-cloud-dataflow itself without interrupting the other python package settings in google datalab? Thanks a lot!

**Comment by chamikaramj** [comment]
"pip install google-cloud-dataflow --upgrade" should upgrade Dataflow to the latest version (0.5.5). We recommend installing Dataflow within a virtual environment.

**Comment by jiajiayb** [comment]
Thank you so much for the response, I just 'pip install google-cloud-dataflow --upgrade' as you suggested and after that I check the 'pip freeze' it indicates the version been updated to 0.5.5. Then when I rerun the same script the error message is as follows:

ImportErrorTraceback (most recent call last)
<ipython-input-1-7e5719b3ba15> in <module>()
     14 # imports
     15 import apache_beam as beam
---> 16 import google.cloud.ml as ml
     17 import google.cloud.ml.analysis as analysis
     18 import google.cloud.ml.io as io

/usr/local/lib/python2.7/dist-packages/google/cloud/ml/__init__.py in <module>()
     14 """Google Cloud Machine Learning SDK.
     15 """
---> 16 from google.cloud.ml.dataflow._analyzer import AnalyzeModel
     17 from google.cloud.ml.dataflow._analyzer import ConfusionMatrix
     18 from google.cloud.ml.dataflow._analyzer import LogLoss

/usr/local/lib/python2.7/dist-packages/google/cloud/ml/dataflow/__init__.py in <module>()
     20 from _analyzer import PrecisionRecall
     21 
---> 22 from _ml_transforms import DeployVersion
     23 from _ml_transforms import Evaluate
     24 from _ml_transforms import Predict

/usr/local/lib/python2.7/dist-packages/google/cloud/ml/dataflow/_ml_transforms.py in <module>()
     25 import apache_beam as beam
     26 from apache_beam import coders as df_coders
---> 27 from apache_beam.utils import options as df_options
     28 import batch_prediction
     29 

ImportError: cannot import name options

I guess this is because there is a conflict between most recent google-cloud-dataflow and other pre-installed packages. Please let me know if you think this is due to something else. Do you have any suggestion on how to fix this error? Any comment will be highly appreciated :) Thanks a lot!

**Comment by charlesccychen** [comment]
@jiajiayb: this may be due to a mismatch between the Cloud ML SDK and the Beam / Dataflow SDK version.  See the instructions here for more details (select the "Local: Mac / Linux" tab of "Setting up your environment"): https://cloud.google.com/ml/docs/how-tos/getting-set-up

You may want to go through the entirety of the instructions, but this command may fix your particular issue:

```
pip install --upgrade --force-reinstall \
    https://storage.googleapis.com/cloud-ml/sdk/cloudml.latest.tar.gz
```

**Comment by jiajiayb** [comment]
@charlesccychen I am sorry for the delayed response. Yes, you are right, the problem was due to the conflict of datalab with newest version of apache beam. The datalab is updating their tutorial and I am expecting the new tutorial will not have the conflict issue mentioned above. Thanks a lot for your help.

**Comment by charlesccychen** [comment]
@jiajiayb Thanks, glad to help!

**Pull request merged**